test/V4Router.t.sol

```diff
@@ -0,0 +1,403 @@
+        token2.mint(address(this), 2 ** 128);
+        token3 = new MockERC20("Test3", "3", 18);
+        token3.mint(address(this), 2 ** 128);
+
+        key0 = createPoolKey(token0, token1, address(0));
+        key1 = createPoolKey(token1, token2, address(0));
+        key2 = createPoolKey(token2, token3, address(0));
+
+        setupPool(key0);
+        setupPool(key1);
+        setupPool(key2);
+
+        token0.approve(address(router), type(uint256).max);
+        token1.approve(address(router), type(uint256).max);
+        token2.approve(address(router), type(uint256).max);
+        token3.approve(address(router), type(uint256).max);
+    }
+
+    function testRouter_bytecodeSize() public {
+        snapSize("RouterBytecode", address(router));
+    }
+
+    function testRouter_swapExactInputSingle_zeroForOne() public {
+        uint256 amountIn = 1 ether;
+        uint256 expectedAmountOut = 992054607780215625;
+
+        IV4Router.ExactInputSingleParams memory params =
+            IV4Router.ExactInputSingleParams(key0, true, address(this), uint128(amountIn), 0, 0, bytes(""));
+
+        uint256 prevBalance0 = token0.balanceOf(address(this));
+        uint256 prevBalance1 = token1.balanceOf(address(this));
+
+        snapStart("RouterExactInputSingle");
+        router.swap(IV4Router.SwapType.ExactInputSingle, abi.encode(params));
+        snapEnd();
+
+        uint256 newBalance0 = token0.balanceOf(address(this));
+        uint256 newBalance1 = token1.balanceOf(address(this));
+
+        assertEq(prevBalance0 - newBalance0, amountIn);
+        assertEq(newBalance1 - prevBalance1, expectedAmountOut);
+    }
+
+    function testRouter_swapExactInputSingle_oneForZero() public {
+        uint256 amountIn = 1 ether;
+        uint256 expectedAmountOut = 992054607780215625;
+
+        IV4Router.ExactInputSingleParams memory params =
+            IV4Router.ExactInputSingleParams(key0, false, address(this), uint128(amountIn), 0, 0, bytes(""));
+
+        uint256 prevBalance0 = token0.balanceOf(address(this));
+        uint256 prevBalance1 = token1.balanceOf(address(this));
+
+        router.swap(IV4Router.SwapType.ExactInputSingle, abi.encode(params));
+
+        uint256 newBalance0 = token0.balanceOf(address(this));
+        uint256 newBalance1 = token1.balanceOf(address(this));
+
+        assertEq(prevBalance1 - newBalance1, amountIn);
+        assertEq(newBalance0 - prevBalance0, expectedAmountOut);
+    }
+
+    function testRouter_swapExactIn_1Hop_zeroForOne() public {
+        uint256 amountIn = 1 ether;
+        uint256 expectedAmountOut = 992054607780215625;
+
+        tokenPath.push(token0);
+        tokenPath.push(token1);
+        IV4Router.ExactInputParams memory params = getExactInputParams(tokenPath, amountIn);
+
+        uint256 prevBalance0 = token0.balanceOf(address(this));
+        uint256 prevBalance1 = token1.balanceOf(address(this));
+
+        snapStart("RouterExactIn1Hop");
```

Comment:
nit: can we refactor the test names to match the tests in core? like `swapExactInOneHop` , and add `_gas` to the end of the test names :)